### ISSUE TITLE
ESLint `@typescript-eslint/no-shadow` error for deserializer of nested Record

### ISSUE THREAD

**Issue by jeremymeng** [Azure/autorest.typescript#3642]
Encounter this in PR https://github.com/Azure/azure-sdk-for-js/pull/36875/files#r2615299169

>/mnt/vss/_work/1/s/sdk/eventgrid/eventgrid-systemevents/src/models/models.ts
    8:1   warning  Unused eslint-disable directive (no problems were reported from '@typescript-eslint/naming-convention')
  461:52  error    'k' is already declared in the upper scope on line 459 column 43                                         @typescript-eslint/no-shadow
  461:55  error    'p' is already declared in the upper scope on line 459 column 46                                         @typescript-eslint/no-shadow
>
>✖ 3 problems (2 errors, 1 warning)
>  0 errors and 1 warning potentially fixable with the `--fix` option.

The generated code is

```ts
/** Information related to the circuit breaker configured on the backend. */
export interface ApiManagementCircuitBreaker {
  /** Overview of all configured rules and respective details. */
  rules: Record<string, Record<string, any>>;
}

export function apiManagementCircuitBreakerDeserializer(item: any): ApiManagementCircuitBreaker {
  return {
    rules: Object.fromEntries(
      Object.entries(item["rules"]).map(([k, p]: [string, any]) => [
        k,
        Object.fromEntries(Object.entries(p).map(([k, p]: [string, any]) => [k, p])),
      ]),
    ),
  };
}
```
